pkg/mapi2capi/aws.go

```diff
@@ -0,0 +1,431 @@
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	warnings = append(warnings, warn...)
+
+	capaMachineTemplate := awsMachineTemplateSpecToAWSMachineTemplate(capaSpec, nil, m.MachineSet.Name, capiNamespace)
+
+	capiMachineSet, warn, err := FromMachineSetToMachineSet(m.MachineSet)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	warnings = append(warnings, warn...)
+
+	// Plug into Core CAPI MachineSet fields that come from the MAPI ProviderConfig which belong here instead of the CAPI AWSMachineTemplate.
+	if awsProviderConfig.Placement.AvailabilityZone != "" {
+		capiMachineSet.Spec.Template.Spec.FailureDomain = ptr.To(awsProviderConfig.Placement.AvailabilityZone)
+	}
+
+	if awsProviderConfig.UserDataSecret != nil && awsProviderConfig.UserDataSecret.Name != "" {
+		capiMachineSet.Spec.Template.Spec.Bootstrap = capiv1.Bootstrap{
+			DataSecretName: &awsProviderConfig.UserDataSecret.Name,
+		}
+	}
+
+	if m.Infrastructure == nil || m.Infrastructure.Status.InfrastructureName == "" {
+		errs = append(errs, errInfrastructureInfrastructureNameCannotBeNil)
+	} else {
+		capiMachineSet.Spec.Template.Spec.ClusterName = m.Infrastructure.Status.InfrastructureName
+		capiMachineSet.Spec.ClusterName = m.Infrastructure.Status.InfrastructureName
+	}
+
+	if len(errs) > 0 {
+		return capiv1.MachineSet{}, capav1.AWSMachineTemplate{}, warnings, utilerrors.NewAggregate(errs)
+	}
+
+	return capiMachineSet, capaMachineTemplate, warnings, nil
+}
+
+// ToMachineTemplateSpec implements the ProviderSpec conversion interface for the AWS provider,
+// it converts AWSProviderSpec to AWSMachineTemplateSpec.
+//
+//nolint:funlen
+func (p AWSProviderSpecAndInfra) ToMachineTemplateSpec() (capav1.AWSMachineTemplateSpec, []string, error) {
+	var (
+		errs     []error
+		warnings []string
+	)
+
+	rootVolume, nonRootVolumes := convertAWSBlockDeviceMappingSpecToCAPI(p.Spec.BlockDevices)
+
+	capiAdditionalSecurityGroups := convertAWSSecurityGroupstoCAPI(p.Spec.SecurityGroups)
+
+	capiAdditionalAWSTags := convertAWSTagsToCAPI(p.Spec.Tags)
+
+	capiIAMInstanceProfile := convertIAMInstanceProfiletoCAPI(p.Spec.IAMInstanceProfile)
+
+	capiMetadataServiceOptions := convertMetadataServiceOptionstoCAPI(p.Spec.MetadataServiceOptions)
+
+	capiSpotMarketOptions := convertAWSSpotMarketOptionsToCAPI(p.Spec.SpotMarketOptions)
+
+	capiAWSResourceReference := convertAWSResourceReferenceToCAPI(p.Spec.Subnet)
+
+	capiAWSAMIReference, err := convertAWSAMIResourceReferenceToCAPI(p.Spec.AMI)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	spec := capav1.AWSMachineTemplateSpec{
+		Template: capav1.AWSMachineTemplateResource{
+			Spec: capav1.AWSMachineSpec{
+				AMI:                      capiAWSAMIReference,
+				AdditionalSecurityGroups: capiAdditionalSecurityGroups,
+				AdditionalTags:           capiAdditionalAWSTags,
+				// CloudInit. Not defined as we use ignition in OpenShift.
```

Comment:
For fields not supported, we should prevent these fields from being present in these resources

Reply:
So adding validation rules for CAPI Machine / CAPA AWSMachine & AWSMachineTemplate or really any CAP* <Provider>Machine and <Provider>MachineTemplate disallowing fields that we don't support/use on the OCP side.

Reply:
Exactly, in the first instance that may be quite restrictive, but we can potentially relax them down the path